server/region_handler.go

```diff
@@ -0,0 +1,496 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math"
+	"net/http"
+	"strconv"
+
+	"github.com/gorilla/mux"
+	"github.com/juju/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/pd/pd-client"
+	"github.com/pingcap/tidb"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/tikv"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/util/codec"
+	goctx "golang.org/x/net/context"
+)
+
+const (
+	pDBName    = "db"
+	pTableName = "table"
+	pRegionID  = "regionID"
+)
+
+const (
+	headerContentType = "Content-Type"
+	contentTypeJSON   = "application/json"
+)
+
+// TableRegions is the response data for list table's regions.
+// It contains regions list for record and indices.
+type TableRegions struct {
+	TableName     string         `json:"name"`
+	TableID       int64          `json:"id"`
+	RecordRegions []RegionMeta   `json:"record_regions"`
+	Indices       []IndexRegions `json:"indices"`
+}
+
+// RegionMeta contains a region's peer detail
+type RegionMeta struct {
+	ID          uint64              `json:"region_id"`
+	Leader      *metapb.Peer        `json:"leader"`
+	Peers       []*metapb.Peer      `json:"peers"`
+	RegionEpoch *metapb.RegionEpoch `json:"region_epoch"`
+}
+
+// IndexRegions is the region info for one index.
+type IndexRegions struct {
+	Name    string       `json:"name"`
+	ID      int64        `json:"id"`
+	Regions []RegionMeta `json:"regions"`
+}
+
+// RegionDetail is the response data for get region by ID
+// it includes indices and records detail in current region.
+type RegionDetail struct {
+	RegionID uint64      `json:"region_id"`
+	StartKey []byte      `json:"start_key"`
+	EndKey   []byte      `json:"end_key"`
+	Frames   []FrameItem `json:"frames"`
+}
+
+// FrameItem includes a index's or record's meta data with table's info.
+type FrameItem struct {
+	DBName    string `json:"db_name"`
+	TableName string `json:"table_name"`
+	TableID   int64  `json:"table_id"`
+	IsRecord  bool   `json:"is_record"`
+	IndexName string `json:"index_name,omitempty"`
+	IndexID   int64  `json:"index_id,omitempty"`
+}
+
+// RegionFrameRange contains a frame range info which the region covered.
+type RegionFrameRange struct {
+	first  *FrameItem        // start frame of the region
+	last   *FrameItem        // end frame of the region
+	region *tikv.KeyLocation // the region
+}
+
+// HTTPRegionHandler is the common field for http region handler. It contains
+// some common functions which would be used in HTTPListTableRegionsHandler and
+// HTTPGetRegionByIDHandler.
+type HTTPRegionHandler struct {
```

Comment:
You can use RegionHandler to handle `/regions/{regionID}` and TableRegionsHandler to handle `/tables/db/table`